src/Evolution/Systems/CurvedScalarWave/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,231 @@
+    const std::optional<tnsr::I<DataVector, Dim, Frame::Inertial>>&
+    /*mesh_velocity*/,
+    const std::optional<Scalar<DataVector>>& normal_dot_mesh_velocity)
+    const noexcept {
+  *packaged_gamma2 = constraint_gamma2;
+  *packaged_interface_unit_normal = interface_unit_normal;
+
+  {  // package characteristic fields
+    const auto char_fields =
+        characteristic_fields(constraint_gamma2, inverse_spatial_metric, psi,
+                              pi, phi, interface_unit_normal);
+    *packaged_v_psi = get<Tags::VPsi>(char_fields);
+    *packaged_v_zero = get<Tags::VZero<Dim>>(char_fields);
+    *packaged_v_plus = get<Tags::VPlus>(char_fields);
+    *packaged_v_minus = get<Tags::VMinus>(char_fields);
+  }
+
+  {  // package characteristic speeds
+    Scalar<DataVector> shift_dot_normal{};
+    get(shift_dot_normal)
+        .set_data_ref(make_not_null(&get<1>(*packaged_char_speeds)));
+    dot_product(make_not_null(&shift_dot_normal), shift, interface_unit_normal);
+    get(shift_dot_normal) *= -1.;
+
+    get<0>(*packaged_char_speeds) =
+        (1. + get(constraint_gamma1)) * get(shift_dot_normal);
+
+    get<2>(*packaged_char_speeds) = get(shift_dot_normal) + get(lapse);
+    get<3>(*packaged_char_speeds) = get(shift_dot_normal) - get(lapse);
+
+    if (normal_dot_mesh_velocity.has_value()) {
+      get<0>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+      get<1>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+      get<2>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+      get<3>(*packaged_char_speeds) -= get(*normal_dot_mesh_velocity);
+    }
+  }
+
+  return max(max(get<0>(*packaged_char_speeds), get<1>(*packaged_char_speeds),
+                 get<2>(*packaged_char_speeds)));
+}
+
+template <size_t Dim>
+void UpwindPenalty<Dim>::dg_boundary_terms(
+    const gsl::not_null<Scalar<DataVector>*> pi_boundary_correction,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        phi_boundary_correction,
+    const gsl::not_null<Scalar<DataVector>*> psi_boundary_correction,
+
+    const Scalar<DataVector>& v_psi_int,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& v_zero_int,
+    const Scalar<DataVector>& v_plus_int, const Scalar<DataVector>& v_minus_int,
+    const Scalar<DataVector>& gamma2_int,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal_int,
+    const tnsr::a<DataVector, 3, Frame::Inertial>& char_speeds_int,
+
+    const Scalar<DataVector>& v_psi_ext,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& v_zero_ext,
+    const Scalar<DataVector>& v_plus_ext, const Scalar<DataVector>& v_minus_ext,
+    const Scalar<DataVector>& gamma2_ext,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>& interface_unit_normal_ext,
+    const tnsr::a<DataVector, 3, Frame::Inertial>& char_speeds_ext,
+    dg::Formulation /*dg_formulation*/) const noexcept {
+  Variables<CurvedScalarWave_detail::char_field_tags<Dim>>
```

Comment:
this function is also pretty brutal on memory allocations. In general, we should do at most 1 memory allocation per interface in `dg_boundary_terms` and one in `dg_package_data`. 

I'll let you and @nikwit coordinate what you think is best here since he's the one using the evolution system. If the conclusion is "let's merge this PR with all the extra allocations and copies" then it would be good to file an issue so that we remember this is a problem

Reply:
Okay, reduce the total number of allocations to 1 buffer allocation. Also reduced the total size of allocation by reusing memory.